### PR TITLE
Replaced `CREATE TEMPORARY TABLE ... LIKE`with different approach

### LINKED ISSUE #37926: [Issue] Replaced `CREATE TEMPORARY TABLE ... LIKE`with different approach
This issue is automatically created based on existing pull request: magento/magento2#36022: Replaced `CREATE TEMPORARY TABLE ... LIKE`with different approach

---------
### Description (*)
Magento's MySQL adapter uses `CREATE TEMPORARY TABLE ... LIKE` command in `\Magento\Framework\DB\Adapter\Pdo\Mysql::createTemporaryTableLike`.

Newer versions of MySQL (8+) have problems with `CREATE TEMPORARY TABLE ... LIKE`, as you can read in the official documentation (https://dev.mysql.com/doc/refman/8.0/en/create-temporary-table.html):

> You cannot use CREATE TEMPORARY TABLE ... LIKE to create an empty table based on the definition of a table that resides in the mysql tablespace, InnoDB system tablespace (innodb_system), or a general tablespace.

The documentation suggests to use `CREATE TEMPORARY TABLE ... SELECT * FROM ... LIMIT 0` instead, but such command **IS NOT** an exact equivalent. It will create temporary table with the same fields, but will not create indexes. Without indexes performance of reindexing big amount of data is absurdly slow.

I suggest to build `CREATE TEMPORARY TABLE` without `LIKE` and without `SELECT * FROM`. Here's my idea:

Firstly use `SHOW CREATE TABLE` to get query for creation of source table (including indexes creation), and then modify it with PHP: change `CREATE TABLE` to `CREATE TEMPORARY TABLE`, add `IF NOT EXISTS` if necessary, remove `TABLESPACE` definition if it exists, and replace source table name with temporary table name.

### Manual testing scenarios (*)
1. Install Magento 2.4 on a server with MySQL 8.
2. Apply sample data.
3. Run `bin/magento index:reindex`.
4. Check data on frontend.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by michalbiarda**
### Description (*)
Magento's MySQL adapter uses `CREATE TEMPORARY TABLE ... LIKE` command in `\Magento\Framework\DB\Adapter\Pdo\Mysql::createTemporaryTableLike`.

Newer versions of MySQL (8+) have problems with `CREATE TEMPORARY TABLE ... LIKE`, as you can read in the official documentation (https://dev.mysql.com/doc/refman/8.0/en/create-temporary-table.html):

> You cannot use CREATE TEMPORARY TABLE ... LIKE to create an empty table based on the definition of a table that resides in the mysql tablespace, InnoDB system tablespace (innodb_system), or a general tablespace.

The documentation suggests to use `CREATE TEMPORARY TABLE ... SELECT * FROM ... LIMIT 0` instead, but such command **IS NOT** an exact equivalent. It will create temporary table with the same fields, but will not create indexes. Without indexes performance of reindexing big amount of data is absurdly slow.

I suggest to build `CREATE TEMPORARY TABLE` without `LIKE` and without `SELECT * FROM`. Here's my idea:

Firstly use `SHOW CREATE TABLE` to get query for creation of source table (including indexes creation), and then modify it with PHP: change `CREATE TABLE` to `CREATE TEMPORARY TABLE`, add `IF NOT EXISTS` if necessary, remove `TABLESPACE` definition if it exists, and replace source table name with temporary table name.

### Manual testing scenarios (*)
1. Install Magento 2.4 on a server with MySQL 8.
2. Apply sample data.
3. Run `bin/magento index:reindex`.
4. Check data on frontend.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#37926: Replaced `CREATE TEMPORARY TABLE ... LIKE`with different approach

### CODE REVIEW

**Comment by michalbiarda** [comment]
@magento give me test instance

**Comment by michalbiarda** [comment]
@magento run all tests

File: lib/internal/Magento/Framework/DB/Adapter/Pdo/Mysql.php
Comment (kandy) [inline lib/internal/Magento/Framework/DB/Adapter/Pdo/Mysql.php]:
```suggestion
        $originCreate = $this->fetchOne("SHOW CREATE TABLE {$originTable}");
```
You can use fetchOne to get result
Also sprintf is unnecessary

Comment (michalbiarda) [reply lib/internal/Magento/Framework/DB/Adapter/Pdo/Mysql.php]: @kandy Unfortunately I can't use `fetchOne` because `SHOW CREATE TABLE` returns two fields. First is table name, the second is SQL query.

I like you suggestion about `sprintf`, though :+1:.

**Comment by michalbiarda** [comment]
@kandy I'm converting this PR to draft. I'll apply your suggestion about `sprintf`, but I will also fix one more thing reported by someone else in the meantime.

**Comment by michalbiarda** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run Static Tests, Functional Tests EE

**Comment by engcom-Hotel** [comment]
Hello @michalbiarda,

Thanks for the contribution!

According to the below comment [here](https://github.com/magento/magento2/pull/36022#issuecomment-1227351474), it seems that you are still working on it.

> @kandy I'm converting this PR to draft. I'll apply your suggestion about `sprintf`, but I will also fix one more thing reported by someone else in the meantime.

Can you please confirm, if the PR is ready for review?

Thanks

**Comment by michalbiarda** [comment]
Hello @engcom-Hotel,

After that comment I added something to my change, and marked this PR as ready for review 10 months ago :sweat_smile:. I confirm it's still ready for review :smile:.

**Comment by engcom-Lima** [comment]
@magento run Functional Tests EE

**Comment by engcom-Hotel** [comment]
Hello @michalbiarda,

Thanks for the contribution!

But we have tried to reproduce the issue in both Magento 2.4.6 and 2.4-develop instance but the issue is not reproducible for us. Please find below the screenshot for reference:

**PHP & MySQL Version**
image

**Product list page after `bin/magento index:reindex` command**
image

Please let us know if we missed anything or if you have any screenshot or screencast of the issue.

Thanks

**Comment by michalbiarda** [comment]
@engcom-Hotel Unfortunately I don't remember exactly how to reproduce the issue (it was 8 months ago), but according to official MySQL 8 documentation (https://dev.mysql.com/doc/refman/8.0/en/create-temporary-table.html):
> You cannot use CREATE TEMPORARY TABLE ... LIKE to create an empty table based on the definition of a table that resides in the mysql tablespace, InnoDB system tablespace (innodb_system), or a general tablespace. The tablespace definition for such a table includes a TABLESPACE attribute that defines the tablespace where the table resides, and the aforementioned tablespaces do not support temporary tables. 

so most probably your test database doesn't use mentioned tablespaces, while the databases at our hosting provider do.

The solution proposed by me doesn't use "LIKE" at all, which prevents the following error during reindexing on MySQL setup with these tablespaces:
> SQLSTATE[HY000]: General error: 1478 InnoDB: Tablespace `innodb_system` cannot contain TEMPORARY tables., query was: CREATE TEMPORARY TABLE IF NOT EXISTS `catalog_category_product_index_store1_tmp` LIKE `catalog_category_product_index_store1`

**Comment by engcom-Hotel** [comment]
Hello @michalbiarda,

Thanks for the reply!

I am referring to the below link:

https://dev.mysql.com/doc/refman/8.0/en/create-temporary-table.html

As per the above link:

> You cannot use CREATE TEMPORARY TABLE ... LIKE to create an empty table based on the definition of a table that resides in the mysql tablespace, InnoDB system tablespace (innodb_system), or a general tablespace. The tablespace definition for such a table includes a TABLESPACE attribute that defines the tablespace where the table resides, and the aforementioned tablespaces do not support temporary tables.


So the query `CREATE TEMPORARY TABLE ... LIKE` will not work for the below Table Space:

- mysql
- innodb_system
- a general tablespace

But as we checked in the database, Magento uses a `File-per-table` tablespace for MySQL tables. Can you please share the scenario or flow where the above-mentioned tablespace has been used in Magento? 

It will help us with the further processing of this PR.

Thanks

**Comment by engcom-Hotel** [comment]
Meanwhile we are moving this PR `On Hold`

**Comment by michalbiarda** [comment]
@engcom-Hotel

Sorry, I was on vacation :smile:.

I think you have to have `innodb_file_per_table=0` in your MySQL config. In such case `innodb_system` tablespace will be the default one used.

`innodb_file_per_table=1` is a default at the moment (I think it was switched to `1` in MySQL 5.6 or 5.7), but if by any reason some database has it set to `0`, they'll get the error I mentioned.

With proposed solution, we don't have to worry about tablespaces at all.

**Review by engcom-Hotel — APPROVED** [review]

**Comment by engcom-Hotel** [comment]
:heavy_check_mark: QA Passed

**Preconditions:**

1. SET `innodb_file_per_table` to `OFF`.

**Manual testing scenario:**

- Install Magento 2.4 on a server with MySQL 8.
- Generate data using the performance toolkit provided by Magento. Please refer to this [link](https://experienceleague.adobe.com/docs/commerce-operations/configuration-guide/cli/generate-data.html?lang=en) for this same.
- Run bin/magento index:reindex.
- Check data on front end.

**Actual Result:** :heavy_check_mark:

Reindex command should run properly without any errors and frontend is accessible with categories and products.

**After:** :heavy_check_mark: 

image

**Before:** :heavy_multiplication_x:

image

Some functional tests are still failing hence moving this PR into extended testing.

Thanks

**Comment by engcom-Echo** [comment]
@magento run Functional Tests EE

**Comment by engcom-Echo** [comment]
Functional Tests EE failures are different on last two run on same commit.
Run1
image

Run2
image

Hence moving to `Merge In progress`

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests EE

**Comment by engcom-Hotel** [comment]
@magento create issue

**Comment by engcom-Hotel** [comment]
Moving back this PR to `Merge in progress`

**Comment by kennylawrence** [comment]
...sooooo..... what exactly is the solution? I currently have to remember to modify the mysql pdo adapter file (vendor/magento/framework/DB/Adapter/Pdo/Mysql.php) every time I update with any new Magento releases. This has been going on for at least four years, ever since MySQL v8. I host at MageMojo/Webscale, and at the moment the following links to their "official" fix for this at the following page whenever I forget and end up creating a new ticket because indexing fails: https://github.com/magemojo/m2-patches/wiki

Is there going to be a more permanent solution? Or is something wrong with the configuration at one of the largest Magento hosting environments (or possibly just my old config is out of date) ? From this issue thread, I wasn't able to grasp what exactly is the current status of the issue (wrong config? code?). Please can we have the final word on this?

**Comment by kennylawrence** [comment]
OK - I just realized that these kinds of fixes aren't added to security releases. I see in 2.4.7 the file has been updated according to the different approach mentioned above, while it does not exist in 2.4.6-p8. I will copy the relevant function over to my file. I am happy this is finally being put to rest!

**Pull request merged**
